hera_cal/redcal.py

```diff
@@ -1957,6 +1963,123 @@ def redcal_run(input_data, filetype='uvh5', firstcal_ext='.first.calfits', omnic
+            delays[ant][tslice] = fc_meta['dlys'][ant]
+            polarity_flips[ant][tslice] = fc_meta['polarity_flips'][ant]
+            offsets[ant][tslice] = fc_meta['offsets'][ant]
+        nt += len(times)
+    # compute medians.
+    for ant in delays:
+        delays[ant] = np.nanmedian(delays[ant])
+        polarity_flips[ant] = bool(np.nanmedian(polarity_flips[ant]))
+        offsets[ant] = np.nanmedian(offsets[ant])
+
+    # update metadata and write out.
+    for filenum, (meta_filename, fc_meta, omni_meta, freqs, times, lsts, antpos, history) in enumerate(redcal_metas):
+        for ant in fc_meta['dlys']:
+            fc_meta['dlys'][ant][:] = delays[ant]
+            fc_meta['polarity_flips'][ant][:] = polarity_flips[ant]
+            fc_meta['offsets'][ant][:] = offsets[ant]
+        save_redcal_meta(meta_filename.replace(output_replace, output_ext), fc_meta, omni_meta, freqs, times, lsts, antpos,
+                         history + '\nTook nightly time median on delay and polarity flips.')
+
+
+def update_redcal_phase_degeneracy(redcal_file, redcal_meta_file, old_redcal_meta_file, output_file=None,
+                                   dont_replace_degens=False, dont_replace_polarities=False,
+                                   clobber=False):
+    """Update the phase degenerate component of a redcal solution and phase flips.
+
+    Parameters
+    ----------
+    redcal_file: str
+        path to calfits file containing redcal solution to update.
+    redcal_meta_file:
+        path to redcal meta file generated with io.save_redcal_meta()  with delays to be used to update degeneracy.
+    old_redcal_meta_file:
+        path to redcal meta file with old polarity flips.
+    output_file: str, optional
+        path to output file. Optional to write output to.
+    replace_degens: bool, optional
+        If True, replace degenerate portion of calibration solution with solution in redcal_meta_file.
+    fix_polarities: bool, optional
+        If True, fix polarities specified in old_redcal_meta_file.
+    clobber: bool, optional
+        overwrite output calfits.
+    """
+    hc = HERACal(redcal_file)
+    # get redundant calibrator.
+    # get reds
+    gains, _, _, _ = hc.read()
+    fc_meta, omni_meta, freqs, times, lsts, antpos, history = read_redcal_meta(redcal_meta_file)
+    fc_meta_old = read_redcal_meta(old_redcal_meta_file)[0]
+    reds = get_reds(antpos, pols=[pol.replace('J', '').replace('j', '') for pol in hc.pols])
+    rc = RedundantCalibrator(reds)
+    firstcal_gains = {}
+
+    if not dont_replace_polarities:
+        for ant in fc_meta['polarity_flips']:
+            gains[ant] *= (-1. + 0j) ** np.abs(fc_meta['polarity_flips'][ant][:, None] - fc_meta_old['polarity_flips'][ant][:, None])
+    ntimes = hc.Ntimes
+    if not dont_replace_degens:
+        for ant in gains:
+            dly_factor = np.exp(2j * np.pi * freqs[None, :] * fc_meta['dlys'][ant][:, None]) * np.exp(1j * fc_meta['offsets'][ant][:, None])
```

Comment:
a bit weird to call it a dly_factor when it includes an offset (which can be summed inside the exponent)